Fix removing non-last list items from Last Update Trigger sensor's settings

marazmarci opened the pull request:
Fixes #2550 _Cannot add new Intent to Last Update Trigger sensor_

## Summary
Now the `Last Update Trigger` sensor's settings screen handles removing non-last list items correctly.

The fix is intended to be backwards compatible, so existing installations with incorrect state will be fixed automatically upon the first call of the `sendLastUpdate()` method (= by an update in the background, or when the user edits the list).

## Screen recordings
[video before the fix: reproduction of the bug](https://user-images.githubusercontent.com/1349654/171994081-0fede94b-c223-46db-93b6-f2c23244b48a.mp4)

[video after the fix: the bug cannot be reproduced anymore](https://user-images.githubusercontent.com/1349654/171994305-86835a33-8b1d-4f8f-a1ff-ada6da129341.mp4)

<!-- ## Link to pull request in Documentation repository -->
<!-- Pull requests that add, change or remove functionality must have a corresponding pull request in the Companion App Documentation repository (https://github.com/home-assistant/companion.home-assistant). Please add the number of this pull request after the "#" -->
<!-- Documentation: home-assistant/companion.home-assistant# -->

## Any other notes
I did some refactoring in the `sendLastUpdate()` method to clarify and optimize things a bit, but I tried to not go too far to keep the changes focused on the bugfix.

Useful command for testing if `Intent`s are received:
`adb shell am broadcast -a com.example.test1`
(don't forget to force close and reopen the app after adding new `Intent`s to the list!)

homeassistant commented:
Hello @marazmarci,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`cc6718e67a33ee818c0dfcf532767633e553c3bf`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;android&#x2F;pull&#x2F;2585&#x2F;commits&#x2F;cc6718e67a33ee818c0dfcf532767633e553c3bf) - This commit has something that looks like an email address (marton.maraz@goto.com). Maybe try linking that to GitHub?.
- [`7a299916871edc45776a520bc7600c7788609961`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;android&#x2F;pull&#x2F;2585&#x2F;commits&#x2F;7a299916871edc45776a520bc7600c7788609961) - This commit has something that looks like an email address (marton.maraz@goto.com). Maybe try linking that to GitHub?.
- [`132c2e9dc9916962372df36a970af2db29ee7bbd`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;android&#x2F;pull&#x2F;2585&#x2F;commits&#x2F;132c2e9dc9916962372df36a970af2db29ee7bbd) - This commit has something that looks like an email address (marton.maraz@goto.com). Maybe try linking that to GitHub?.
- [`ee4a8ea73c4f4789fa5a07eabb31669b861b6e7b`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;android&#x2F;pull&#x2F;2585&#x2F;commits&#x2F;ee4a8ea73c4f4789fa5a07eabb31669b861b6e7b) - This commit has something that looks like an email address (marton.maraz@goto.com). Maybe try linking that to GitHub?.
- [`51f7736620f9bdd9fb013a0b211dc7713d8b2e4f`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;android&#x2F;pull&#x2F;2585&#x2F;commits&#x2F;51f7736620f9bdd9fb013a0b211dc7713d8b2e4f) - This commit has something that looks like an email address (marton.maraz@goto.com). Maybe try linking that to GitHub?.
- [`6d44bd9f32ff6cb7df408f80da600746d3d549a9`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;android&#x2F;pull&#x2F;2585&#x2F;commits&#x2F;6d44bd9f32ff6cb7df408f80da600746d3d549a9) - This commit has something that looks like an email address (marton.maraz@goto.com). Maybe try linking that to GitHub?.
- [`1d0776daf1dda01051cc5e496b603e8aa1cb4612`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;android&#x2F;pull&#x2F;2585&#x2F;commits&#x2F;1d0776daf1dda01051cc5e496b603e8aa1cb4612) - This commit has something that looks like an email address (marton.maraz@goto.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

marazmarci commented:
Whoops, committed with the wrong git user. Fixed!

jpelgrom commented on common/src/main/java/io/homeassistant/companion/android/common/sensors/LastUpdateManager.kt:
```diff
@@ -64,21 +65,49 @@ class LastUpdateManager : SensorManager {
+        }
+        val shouldAddNewIntent = allSettings.firstOrNull { it.name == SETTING_ADD_NEW_INTENT }?.value == "true"
+        if (shouldAddNewIntent) {
+            val newIntentSettingOrdinal = currentHighestIntentSettingOrdinal + 1
```
`currentHighestIntentSettingOrdinal` isn't updated if the settings are renumbered, which might give a number that is too high instead of the expected value. For example: if there were settings for 1, 3, 4, it will be renumbered to 1, 2, 3, so the expected new setting should be 4 but it will be 5 (`currentHighestIntentSettingOrdinal` = 4).

jpelgrom commented on common/src/main/java/io/homeassistant/companion/android/common/sensors/LastUpdateManager.kt:
```diff
@@ -64,21 +65,49 @@ class LastUpdateManager : SensorManager {
+            val newIntentSettingOrdinal = currentHighestIntentSettingOrdinal + 1
+            val newIntentSettingName = "$INTENT_SETTING_PREFIX$newIntentSettingOrdinal:"
+            val intentSettingAlreadyExists = allSettings.any { it.name == newIntentSettingName }
+            check(!intentSettingAlreadyExists)
```
I can't think of a scenario where this happens, but if you decide to keep it, instead of failing this should be a recoverable error.

marazmarci commented:
@jpelgrom, if you have the necessary rights, could you please remove the `cla-error` label from the PR? It's not relevant anymore.

marazmarci commented:
I marked this PR as `[WIP]`, because it's waiting for this PR to be merged: #2593 _Update delete methods in daos_

(see [this](https://github.com/home-assistant/android/pull/2593/files#r892565540) comment for more information)

marazmarci commented:
Now as #2593 is merged, I updated this PR to use the new `SensorDao::removeSettings(...)` method and rebased onto `master`.

@jpelgrom, could you please re-review it?

For simplicity, here's the diff between the current and the original branch before the rebase:
![Home Assistant Android GitHub Pull Request - Fix removing non-last list items from Last Update Trigger sensor's settings - compare before and after rebase](https://user-images.githubusercontent.com/1349654/173690992-8bc8c4c7-315c-4eab-9b2e-4328f5ab7f58.png)

jpelgrom commented on common/src/main/java/io/homeassistant/companion/android/common/sensors/LastUpdateManager.kt:
```diff
@@ -64,24 +65,40 @@ class LastUpdateManager : SensorManager {
+        if (shouldAddNewIntent) {
+            val newIntentSettingOrdinal = intentSettings.size + 1
+            val newIntentSettingName = "$INTENT_SETTING_PREFIX$newIntentSettingOrdinal:"
+            val intentSettingAlreadyExists = allSettings.any { it.name == newIntentSettingName }
+            if (!intentSettingAlreadyExists) {
```
Suggestion: use `allSettings.none` to simplify the code and remove the variable

marazmarci replied:
Done. 4932b60
